crates/chain-config/src/config/codec/encoder.rs

```diff
@@ -0,0 +1,304 @@
+use std::{
+    fs::File,
+    path::{
+        Path,
+        PathBuf,
+    },
+};
+
+use ::parquet::basic::{
+    Compression,
+    GzipLevel,
+};
+
+use crate::{
+    config::{
+        codec::parquet::Schema,
+        contract_balance::ContractBalance,
+        contract_state::ContractStateConfig,
+    },
+    CoinConfig,
+    ContractConfig,
+    MessageConfig,
+    StateConfig,
+};
+
+use super::parquet;
+
+enum EncoderType {
+    Json {
+        buffer: StateConfig,
+        state_file_path: PathBuf,
+    },
+    Parquet {
+        coins: parquet::Encoder<File, CoinConfig>,
+        messages: parquet::Encoder<File, MessageConfig>,
+        contracts: parquet::Encoder<File, ContractConfig>,
+        contract_state: parquet::Encoder<File, ContractStateConfig>,
+        contract_balance: parquet::Encoder<File, ContractBalance>,
+    },
+}
+
+pub struct Encoder {
+    encoder: EncoderType,
+}
+
+impl Encoder {
+    pub fn json(snapshot_dir: impl AsRef<Path>) -> Self {
+        Self {
+            encoder: EncoderType::Json {
+                state_file_path: snapshot_dir.as_ref().join("state.json"),
+                buffer: StateConfig::default(),
+            },
+        }
+    }
+
+    pub fn parquet(
+        snapshot_dir: impl AsRef<Path>,
+        compression_level: u8,
+    ) -> anyhow::Result<Self> {
+        fn create_encoder<T>(
+            path: &Path,
+            name: &str,
+            compression: Compression,
+        ) -> anyhow::Result<parquet::Encoder<File, T>>
+        where
+            T: Schema,
+        {
+            let path = path.join(format!("{name}.parquet"));
+            let file = std::fs::File::create(path)?;
+            parquet::Encoder::new(file, compression)
+        }
+
+        let path = snapshot_dir.as_ref();
+        let compression =
```

Comment:
Do we need to know `compression` during decoding?

Reply:
It is read automatically from the metadata inside the parquet files.